durable-storage/src/persistence_layer.rs

```diff
@@ -0,0 +1,170 @@
+// SPDX-FileCopyrightText: 2025 Trilitech <contact@trili.tech>
+//
+// SPDX-License-Identifier: MIT
+
+//! Implementation of the RISC-V PVM durable storage's persistence layer.
+//!
+//! A persistence layer is tied to a repository on disk, identified by a directory.
+//! Within that directory, the persistence layer needs to be able to perform basic KV operations:
+//! - get
+//! - set
+//! - delete
+//!
+//! As well as repository-level operations:
+//! - new
+//! - clone
+//! - commit (returning the commit hash)
+//! - checkout a specific commit
+//!
+//! The folder structure of the [`DirectoryManager`] is:
+//! ```
+//! <repo_path>:
+//!    temporary/
+//!        db_<random>/checkpoint/
+//!            <rocksdb internals>
+//! ```
+
+use std::mem::ManuallyDrop;
+
+use crate::repo::DirectoryManager;
+use crate::repo::DirectoryManagerError;
+
+/// Type alias for a 32-byte hash used for identifying key-value blobs & commits.
+type Hash = [u8; 32];
+
+/// Errors encountered when interacting with the persistence layer.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("RocksDB error: {0}")]
+    RocksDB(#[from] rocksdb::Error),
+
+    #[error("Directory manager error: {0}")]
+    DirectoryManager(#[from] DirectoryManagerError),
+}
+
+/// Create a [`rocksdb::Options`] instance for creating a new database.
+fn options_expect_newly_created() -> rocksdb::Options {
+    // To avoid accidentaly overwriting an existing database, we set `error_if_exists`.
+    let mut options = rocksdb::Options::default();
+    options.create_if_missing(true);
+    options.set_error_if_exists(true);
+    options
+}
+
+/// Mode in which the [`PersistenceLayer`] was instantiated.
+///
+/// - `ReadOnly` corresponds to a database checked out from a specific commit.
```

Comment:
This seems to imply you wouldn't be able to create a new commit starting from a checked-out commit?

Reply:
Changed [here](https://github.com/tezos/riscv-pvm/compare/b808d5a3497686deaa91c452eada08de6f24398b..789a1f0d182e1079cacf1a08e6a33eb4c48c3660)